datasource/ethereum/src/ethereum_adapter.rs

```diff
@@ -579,26 +652,199 @@ where
+                    call_filter_opt.unwrap(),
+                )));
+            }
+            if block_filter_opt.is_some() {
+                let block_filter = block_filter_opt.unwrap();
+                match block_filter.contract_addresses.len() {
+                    0 => (),
+                    _ => {
+                        let call_filter = EthereumCallFilter::from(block_filter);
+                        block_futs.push(Box::new(eth.blocks_with_calls(
+                            &logger,
+                            from,
+                            to,
+                            call_filter,
+                        )));
+                    }
+                }
+            }
+        }
+        Box::new(
+            future::join_all(block_futs).and_then(|block_pointer_chunks| {
+                let mut blocks = block_pointer_chunks
+                    .into_iter()
+                    .flatten()
+                    .collect::<Vec<EthereumBlockPointer>>();
+                blocks.sort_by(|a, b| a.number.cmp(&b.number));
+                // Dedup only remove consecutive duplicates, so it needs to be
+                // run after the vector is sorted
+                blocks.dedup();
+                future::ok(blocks)
+            }),
+        )
+    }
+
+    fn blocks(
+        &self,
+        logger: &Logger,
+        from: u64,
+        to: u64,
+    ) -> Box<Future<Item = Vec<EthereumBlockPointer>, Error = Error> + Send> {
+        let eth = self.clone();
+        let logger = logger.clone();
+
+        // Generate `EthereumBlockPointers` from `to` backwards to `from`
+        let block_ptrs = self
+            .block_hash_by_block_number(&logger, to)
+            .map(move |block_hash_opt| EthereumBlockPointer {
+                hash: block_hash_opt.unwrap(),
+                number: to,
+            })
+            .and_then(move |block_pointer| {
+                stream::unfold(block_pointer, move |descendant_block_pointer| {
+                    if descendant_block_pointer.number < from {
+                        return None;
+                    }
+                    // Populate the parent block pointer
+                    Some(
+                        eth.block_parent_hash_by_block_hash(&logger, descendant_block_pointer.hash)
+                            .map(move |block_hash_opt| {
+                                let parent_block_pointer = EthereumBlockPointer {
+                                    hash: block_hash_opt.unwrap(),
+                                    number: descendant_block_pointer.number - 1,
+                                };
+                                (descendant_block_pointer, parent_block_pointer)
+                            }),
+                    )
+                })
+                .collect()
+            })
+            .map(move |mut block_pointers| {
+                block_pointers.reverse();
+                block_pointers
+            });
+        Box::new(block_ptrs)
```

Comment:
I'd drop the `let block_ptrs` assignment and instead do `Box::new(...)` right away. The fewer variables, the fewer things can be mutated. 😉